### PR TITLE
fix: read and write block in parallel when compact

### LINKED ISSUE #8878: bug: COMPACT is slow and cpu load is low
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v0.8.117-nightly

### What's Wrong?

Compact execution is slow and cpu load is low.

image
image
block information:
image
image

### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Read and write block in parallel when compact transform.

1. ReadBlocks(async):  read blocks concurrently. set max_memory to limit the size of block memory to be read.
2. CompactBlocks(sync): execute compact operation sequentially on the fetched blocks, and generate the new index_data and BlockMeta.
3. SerializedBlocks(async):  write block data and index data concurrently.

Closes #8878

### CODE REVIEW

File: src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs [hunk 1]
```diff
@@ -143,18 +184,96 @@ impl Processor for CompactTransform {
...
+        let meta = input_date.get_meta().unwrap();
+        let task_meta = CompactSourceMeta::from_meta(meta)?;
+        self.order = task_meta.order;
+        self.compact_tasks = task_meta.tasks.clone();
+
+        self.state = State::ReadBlocks;
         Ok(Event::Async)
     }
 
     fn process(&mut self) -> Result<()> {
```
Comment (BohuTANG) [inline src/query/storages/fuse/fuse/src/operations/mutation/compact/compact_transform.rs:197]: Could you add more docs for the parallel compact algorithm?
It's better to read the doc than the code.

**Review by zhang2014 — APPROVED** [review]

**Comment by zhyass** [comment]
Find a bug:
```
mysql> create table t(c int) block_per_segment=4 row_per_block=3;
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values(1);
Query OK, 0 rows affected (0.03 sec)

mysql> insert into t values(2);
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t values(3);
Query OK, 0 rows affected (0.03 sec)

mysql> insert into t values(4);
Query OK, 0 rows affected (0.03 sec)

mysql> select * from t;
+------+
| c    |
+------+
|    4 |
|    3 |
|    2 |
|    1 |
+------+
4 rows in set (0.02 sec)
Read 4 rows, 16.00 B in 0.010 sec., 416.94 rows/sec., 1.63 KiB/sec.

mysql> optimize table t compact;
Query OK, 0 rows affected (0.03 sec)

mysql> select * from t;
Empty set (0.02 sec)
Read 0 rows, 0.00 B in 0.004 sec., 0 rows/sec., 0.00 B/sec.
```

**Comment by BohuTANG** [comment]
Please ignore `max_memory_usage `, we have not done the quota.

**Review by BohuTANG — APPROVED** [review]

**Review by dantengsky — APPROVED** [review]
LGTM

**Pull request merged**